gateway-js/src/dataRewrites.ts

```diff
@@ -0,0 +1,130 @@
+import { FetchDataRewrite } from "@apollo/query-planner";
+import { assert } from "console";
+import { GraphQLSchema, isAbstractType, isInterfaceType, isObjectType } from "graphql";
+
+const FRAGMENT_PREFIX = '... on ';
+
+export function applyRewrites(schema: GraphQLSchema, rewrites: FetchDataRewrite[] | undefined,  value: Record<string, any>) {
+  if (!rewrites) {
+    return;
+  }
+
+  for (const rewrite of rewrites) {
+    applyRewrite(schema, rewrite, value);
+  }
+}
+
+function applyRewrite(schema: GraphQLSchema, rewrite: FetchDataRewrite,  value: Record<string, any>) {
+  const splitted = splitPathLastElement(rewrite.path);
+  if (!splitted) {
+    return;
+  }
+
+  const [parent, last] = splitted;
+  const { kind, value: fieldName } = parsePathElement(last);
+  // So far, all rewrites finish by a field name. If this ever changes, this assertion will catch it early and we can update.
+  assert(kind === 'fieldName', () => `Unexpected fragment as last element of ${rewrite.path}`);
+  applyAtPath(schema, parent, value, rewriteAtPathFunction(rewrite, fieldName));
+}
+
+function rewriteAtPathFunction(rewrite: FetchDataRewrite, fieldAtPath: string): (obj: Record<string, any>) => void {
+  switch (rewrite.kind) {
+    case 'ValueSetter':
+      return (obj) => {
+        obj[fieldAtPath] = rewrite.setValueTo;
+      };
+    case 'KeyRenamer':
+      return (obj) => {
+        obj[rewrite.renameKeyTo] = obj[fieldAtPath];
+        obj[fieldAtPath] = undefined;
+      };
+  }
+}
+
+
+/**
+ * Given a path, separates the last element of path and the rest of it and return them as a pair.
+ * This will return `undefined` if the path is empty.
+ */
+function splitPathLastElement(path: string[]): [string[], string] | undefined {
+  if (path.length === 0) {
+    return undefined;
+  }
+
+  const lastIdx = path.length - 1;
+  return [path.slice(0, lastIdx), path[lastIdx]];
+}
+
+function applyAtPath(schema: GraphQLSchema, path: string[], value: any, fct: (objAtPath: Record<string, any>) => void) {
+  if (Array.isArray(value)) {
+    for (const arrayValue of value) {
+      applyAtPath(schema, path, arrayValue, fct);
+    }
+    return;
+  }
+
+  if (typeof value !== 'object') {
```

Comment:
I'm a little confused. Further down, we seem to be treating `value` as a `Record<string,any>`. Wouldn't this cause that case to exit here? I'm probably missing something.

Reply:
If value is a `Record<string, any>`, then it is an `object`, so it shouldn't exit since the condition exits if it is _not_ an object.

Reply:
If value is a `Record<string, any>`, then it is an `object`, so it shouldn't exit since the condition exits if it is _not_ an object.